src/merkle.rs

```diff
@@ -567,48 +1134,73 @@ impl<T: Element, A: Algorithm<T>, K: Store<T>, U: Unsigned> MerkleTree<T, A, K,
 
     /// Returns data reference
     #[inline]
-    pub fn data(&self) -> &K {
-        &self.data
-    }
-
-    /// Returns mutable data reference
-    #[inline]
-    pub fn data_mut(&mut self) -> &mut K {
-        &mut self.data
+    pub fn data(&self) -> &S {
+        &self.data.store().unwrap()
     }
 
     /// Returns merkle leaf at index i
     #[inline]
-    pub fn read_at(&self, i: usize) -> Result<T> {
-        self.data.read_at(i)
+    pub fn read_at(&self, i: usize) -> Result<E> {
+        if TopTreeArity::to_usize() > 0 {
+            // Locate the top-layer tree the sub-tree leaf is contained in.
+            ensure!(
+                TopTreeArity::to_usize() == self.data.sub_trees().unwrap().len(),
+                "Top layer tree shape mis-match"
```

Comment:
Now that you have the enum, I believe instead of doing the tree arity check + unwrap you could do sth like this

```rust
match self.data {
  Data::Variant1 { ref sub_trees } => {
    // do thing
  },
  Data::Variant2 { ref base_trees } => {
    // do other thing
  }
  Data::Variant3 { ref data } => {
      data.store().read_at(i)
  } 
}
```

Reply:
This was one of my main motiviations to suggest this internal change, as it makes it much easier to follow these methods that work against the different variants.